Closes #239: Add SubFlow state repeat ability (#243)
serverlessworkflow/specification: specification.md

Each referenced workflow receives the SubFlow states data as workflow data input.

The `waitForCompletion` property defines if the SubFlow state should wait until the referenced reusable workflow
has completed its execution. If it's set to "true" (default value), SubFlow state execution must wait until the referenced workflow has completed its execution.
In this case the workflow data output of the referenced workflow can and should be merged with the SubFlow states state data.
If it's set to "false" the parent workflow can continue its execution while the referenced sub-workflow 
is being executed. For this case, the referenced (child) workflow data output cannot be merged with the SubFlow states
state data (as by the time its completion the parent workflow execution has already continued).

> The `repeat` property defines the SubFlow states repeated execution (looping) behavior. This allows you to specify that 
> the sub-workflow should be executed multiple times repeatedly.
> If the `repeat` property is defined, the `waitForCompletion` should be assumed have the value of `true`.
> If the workflow explicitly triggers [compensation](#Workflow Compensation) and the SubFlow state 
> was executed and defines its compensation state, it should be compensated once, no matter how many times
> its exection was executed as defined by the `repeat` property.
> After each execution of the SubFlow state, if `repeat` is defined, the SubFlow state data at the end of the 
> one execution should become the state data of the next execution.

Reviewer: ```suggestion
The `repeat` property defines the SubFlow states repeated execution (looping) behavior. This allows you to specify that 
the sub-workflow should be executed multiple times repeatedly.
If the `repeat` property is defined, the `waitForCompletion` should be assumed to have the value of `true`.
If the workflow explicitly triggers [compensation](#Workflow Compensation) and the SubFlow state 
was executed and defines its compensation state, it should be compensated once, no matter how many times
its execution was executed as defined by the `repeat` property.
After each execution of the SubFlow state, if `repeat` is defined, the SubFlow state data at the end of the 
one execution should become the state data of the next execution.
```

| continueOnError | If set to `true` repeats executions in a case unhandled errors propagate from the sub-workflow to this state | boolean | no |
| stopOnEvents | List referencing defined consumed workflow events. SubFlow will repeat execution until one of the defined events is consumed, or until the max property count is reached | array | no |

<details><summary><strong>Click to view example definition</strong></summary>
<p>

<table>
<tr>
    <th>JSON</th>
    <th>YAML</th>
</tr>
<tr>
<td valign="top">

```json
{
  "max": 10,
  "continueOnError": true
}
```

</td>
<td valign="top">

```yaml
max: 10
continueOnError: true
```

</td>
</tr>
</table>

</details>

Repeat definition can be used in [SubFlow](#SubFlow-State) states to define repeated execution (looping).

The `expression` parameter is a [workflow expression](#Workflow-Expressions). It is 
evaluated against SubFlow states data. 
SubFlow state should repeat its execution as long as this expression evaluates to `true` (the expression returns a non-empty result),
or until the `max` property limit is reached. This parameter allows you to stop repeat execution based on data.

> The `checkBefore` property can be used to decide if the `expression` evaluation should be done before or after 
> each SubFlow state execution. Default value of this property is `true`.
> 
> The `max` property sets the maximum count of repeat executions. It should be a positive integer value.
> Runtime implementations must define an internal repeat/loop counter which is incremented for each of the 
> SubFlow state repeated executions. If this counter reaches the max value, repeated executions should end.
> 
> The `continueOnError` property defines if repeated executions should continue or not in case unhandled errors are propagated
> by the sub-workflow to the SubFlow state. Default value of this property is `false`.
> Unhandled errors are errors which are not explicitly handled by the sub-workflow, and the SubFlow state 
> via its [`onErrors`](##Error-Definition) definition.
> 
> If `continueOnError` is set to `false` (default value), and an unhandled error occurs, it should be handled 
> as any other unhandled workflow error, meaning repeat execution shall stop and workflow should stop its exception.
> 
> If an error occurs which propagates to the SubFlow state, and is handled explicitly by the 
> SubFlow states [`onErrors`](#Error-Definition) definition, the control flow must take the path of the error handling definition
> and repeat execution must halt.
> 
> An alternative way to limit repeat executions is via the `stopOnEvents` property. It contains a list of one or more 
> defined consumed workflow events (referened by the unique event name). When `stopOnEvents` is defined,
> SubFlow will repeat execution until one of the defined events is consumed, or until the max property count is reached.

Reviewer: ```suggestion
The `checkBefore` property can be used to decide if the `expression` evaluation should be done before or after each SubFlow state execution. The default value of this property is `true`.

The `max` property sets the maximum count of repeat executions. It should be a positive integer value.
Runtime implementations must define an internal repeat/loop counter, which is incremented for each of the 
SubFlow state repeated executions. If this counter reaches the max value, repeated executions should end.

The `continueOnError` property defines if repeated executions should continue or not if the sub-workflow propagates unhandled errors to the SubFlow state. The default value of this property is `false`.
Unhandled errors are errors that are not explicitly handled by the sub-workflow, and the SubFlow state 
via its [`onErrors`](##Error-Definition) definition.

If `continueOnError` is set to `false` (default value), and an unhandled error occurs, it should be handled as any other unhandled workflow error, meaning repeat execution shall stop and workflow should stop its exception.

If an error occurs which propagates to the SubFlow state and is handled explicitly by the 
SubFlow states [`onErrors`](#Error-Definition) definition, the control flow must take the path of the error handling definition, and repeat execution must halt.

An alternative way to limit repeat executions is via the `stopOnEvents` property. It contains a list of defined consumed workflow events (referenced by the unique event name). When `stopOnEvents` is defined, SubFlow will repeat execution until one of the defined events is consumed or until the max property count is reached.
```
